net: lwm2m: Add functional tests for LwM2M against Leshan

SeppoTakalo opened the pull request:
Use the new Pytest integration to run testcases against Leshan.

Test with:
twister -T tests/net/lib/lwm2m/functional -p native_posix -vv

This requires Leshan running in localhost using following setup: tcp/8080 Leshan web interface and REST API
tcp/8081 Leshan bootstrap server REST API
udp/5683 Leshan non-secure CoAP
udp/5684 Leshan DTLS CoAP
udp/5783 non-secure Bootstrap CoAP
udp/5684 DTLS Bootstrap CoAP

Leshan and Boostrap server must be reachable from native_posix run using IP address 192.0.2.2 as in most of the examples.

SeppoTakalo committed:
net: lwm2m: Add functional tests for LwM2M against Leshan

Use the new Pytest integration to run testcases against Leshan.

Test with:
twister -T tests/net/lib/lwm2m/functional -p native_posix -vv

This requires Leshan running in localhost using following setup:
tcp/8080 Leshan web interface and REST API
tcp/8081 Leshan bootstrap server REST API
udp/5683 Leshan non-secure CoAP
udp/5684 Leshan DTLS CoAP
udp/5783 non-secure Bootstrap CoAP
udp/5684 DTLS Bootstrap CoAP

Leshan and Boostrap server must be reachable from native_posix
run using IP address 192.0.2.2 as in most of the examples.

Tests are written from test spec;
OMA Enabler Test Specification (Interoperability) for
Lightweight M2M

Following tests are implemented in this commit:
* LightweightM2M-1.1-int-0 – Client Initiated Bootstrap
* LightweightM2M-1.1-int-1 – Client Initiated Bootstrap Full (PSK)
* LightweightM2M-1.1-int-101 – Initial Registration
* LightweightM2M-1.1-int-102 – Registration Update
* LightweightM2M-1.1-int-104 – Registration Update Trigge
* LightweightM2M-1.1-int-105 - Discarded Register Update
* LightweightM2M-1.1-int-107 – Extending the lifetime of a registration
* LightweightM2M-1.1-int-108 – Turn on Queue Mode
* LightweightM2M-1.1-int-109 – Behavior in Queue Mode
* LightweightM2M-1.1-int-201 – Querying basic information in Plain Text
* LightweightM2M-1.1-int-203 – Querying basic information in TLV format
* LightweightM2M-1.1-int-204 – Querying basic information in JSON format
* LightweightM2M-1.1-int-205 – Setting basic information in Plain Text
* LightweightM2M-1.1-int-211 – Querying basic information in CBOR format
* LightweightM2M-1.1-int-212 – Setting basic information in CBOR format
* LightweightM2M-1.1-int-215 – Setting basic information in TLV format
* LightweightM2M-1.1-int-220 – Setting basic information in JSON format
* LightweightM2M-1.1-int-221 – Attempt to perform operations on Security
* LightweightM2M-1.1-int-401 – UDP Channel Security – PSK Mode

Signed-off-by: Seppo Takalo <seppo.takalo@nordicsemi.no>

SeppoTakalo commented:
@PerMac You might want to consider if that `Shell` module should be part of the Pytest plugins. Now it is inside a test cases, but I think it will be usefull for those who utilize the shell interface.

PerMac commented:
Do you mean a fixture standardizing the communication with shell, like a one your adding? I think it is a good idea. After the code freeze we want to extract common functions and functions useful during testing to a separate, reusable module. We definitely don't want to end in a situation where each tests defines its own look_for_regex, wait_for_prompt, etc. methods. Having users using such methods with clear interface will also allow us to improve the methods themselves to match changes in pytest adapters, whiteout rewriting each test.

gchwier commented on tests/net/lib/lwm2m/functional/pytest/test_lwm2m.py:
```diff
@@ -0,0 +1,232 @@
+logger = logging.getLogger(__name__)
+
+class Shell(object):
+    def __init__(self, dut) -> None:
```
You may use annotations (`dut: DeviceAbstract`). It will be much easier to read/develop the code with some IDE

gchwier commented on tests/net/lib/lwm2m/functional/pytest/test_lwm2m.py:
```diff
@@ -0,0 +1,232 @@
+            line = next(self._dut.iter_stdout)
+            if line is None or len(line) == 0:
+                continue
+            print(line)
```
Use `logger.debug` instead of print

gchwier commented on tests/net/lib/lwm2m/functional/pytest/test_lwm2m.py:
```diff
@@ -0,0 +1,232 @@
+        timeout = 5  # [sec]
+        time_end = time.time() + timeout
+        while time.time() < time_end:
+            line = next(self._dut.iter_stdout)
```
Now you can find some problems when calling `iter_stdout` multiple times (it is a Generator). This PR #58491 should fix it

gopiotr commented:
> PerMac You might want to consider if that `Shell` module should be part of the Pytest plugins. Now it is inside a test cases, but I think it will be usefull for those who utilize the shell interface.

@SeppoTakalo in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/61224 I added helper `Shell` class, which could be used in your test:
https://github.com/gopiotr/zephyr/blob/pr/align_adapters_API/scripts/pylib/pytest-twister-harness/src/twister_harness/helpers/shell.py
An example of usage of this helper `Shell` class is placed here:
https://github.com/gopiotr/zephyr/blob/pr/align_adapters_API/samples/subsys/testsuite/pytest/shell/pytest/test_shell.py

I would be grateful if you could give some feedback and tell if this is enough and if it works for you.

SeppoTakalo commented:
* Rebased on top of main and uses the new Shell wrapper class.
* Submitted a PR https://github.com/zephyrproject-rtos/zephyr/pull/62760 to help reading values from shell.
* Added few more tests, notably TLV content format tests.

SeppoTakalo commented:
I'm converting this to a proper PR because I think this has reached a point that it is useful and all the required functionality is merged in Zephyr already.

It is up to a debate how we should handle the dependency to a running Leshan server instance. Is it enough to just document it, or should I try to modify/generate a Docker image that someone might run it against?

rlubos commented on tests/net/lib/lwm2m/functional/testcase.yaml:
```diff
@@ -0,0 +1,14 @@
+tests:
+  sample.net.lwm2m:
```
Looks like a copy-paste.

rlubos commented on tests/net/lib/lwm2m/functional/CMakeLists.txt:
```diff
@@ -0,0 +1,10 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(lwm2m_client)
```
Should we use a more descriptive name for the project, like `lwm2m_interop_tests`?

rlubos reviewed (commented):
It looks that all of the dependency commits are already on main, the PR could be rebased?

> It is up to a debate how we should handle the dependency to a running Leshan server instance. Is it enough to just document it, or should I try to modify/generate a Docker image that someone might run it against?

There already is a docker image for networking testing from the old days (see https://github.com/zephyrproject-rtos/zephyr/issues/29764), perhaps it could be updated with Leshan and used for this purpose? Maybe @jukkar could provide more information.

PerMac reviewed (changes requested):
Since this test doesn't work out-of-the box, please add:
- a short README describing how to execute this tests (how to setup environment, start server, etc.)
- a mechanism to mark the test as skipped by pytest if the requirements are not fulfilled. The CI is green on these tests, because they are marked as "slow" and such tests are not executed by default. However, with `--enable-slow` a user can try executing the test and it will fail due to lacking pieces. Your tests can verify if leshan and CoAPthon3 are available in the user's env and if not mark themselves as skipped with a proper reason

gchwier reviewed (dismissed):
looks good, just minor comments

SeppoTakalo commented on tests/net/lib/lwm2m/interop/pytest/test_lwm2m.py:
```diff
@@ -0,0 +1,449 @@
+        #
+
+        # Write bootsrap server information and PSK keys
+        shell.exec_command('lwm2m write 0/0/0 -s coaps://192.0.2.2:5784')
```
Thanks.
Added
```py
LESHAN_IP: str = '192.0.2.2'
COAP_PORT: int = 5683
COAPS_PORT: int = 5684
BOOTSTRAP_COAPS_PORT: int = 5784
```

SeppoTakalo commented on tests/net/lib/lwm2m/interop/testcase.yaml:
```diff
@@ -0,0 +1,14 @@
+      - native_posix
+    platform_allow:
+      - native_posix
+      - qemu_cortex_m3
```
Removed. The framework and shell does not yet work with X86 event that the application would.

SeppoTakalo commented on tests/net/lib/lwm2m/interop/pytest/test_lwm2m.py:
```diff
@@ -0,0 +1,449 @@
+
+def verify_LightweightM2M_1_1_int_0(shell):
+    logger.info("LightweightM2M-1.1-int-0 – Client Initiated Bootstrap")
+    shell._device.readlines_until(regex = '.*Bootstrap started with endpoint', timeout = 5.0)
```
Thanks. Fixed.

SeppoTakalo commented on tests/net/lib/lwm2m/interop/pytest/test_lwm2m.py:
```diff
@@ -0,0 +1,449 @@
+#
+
+def verify_LightweightM2M_1_1_int_0(shell):
+    logger.info("LightweightM2M-1.1-int-0 – Client Initiated Bootstrap")
```
Thanks. These seem to have come from copy&pasting from the test-spec PDF.
I replaced with ` - `

SeppoTakalo commented on tests/net/lib/lwm2m/interop/pytest/leshan.py:
```diff
@@ -0,0 +1,104 @@
+            resp = self.get('/security/clients')
+            if not isinstance(resp, list):
+                raise RuntimeError('Did not receive list of endpoints')
+        except requests.exceptions.ConnectionError as e:
```
Fixed.

SeppoTakalo commented on tests/net/lib/lwm2m/interop/pytest/test_lwm2m.py:
```diff
@@ -0,0 +1,449 @@
+# https://www.openmobilealliance.org/release/LightweightM2M/ETS/OMA-ETS-LightweightM2M-V1_1-20190912-D.pdf
+#
+
+def verify_LightweightM2M_1_1_int_0(shell):
```
Thanks. Annotated
```py
def verify_LightweightM2M_1_1_int_104(shell: Shell, leshan: Leshan, endpoint: str):
```

SeppoTakalo commented on tests/net/lib/lwm2m/interop/pytest/test_lwm2m.py:
```diff
@@ -0,0 +1,449 @@
+    return HelperClient(server=('127.0.0.1', 5683))
+
+@pytest.fixture(scope='function')
+def shell(dut: DeviceAdapter) -> Shell:
```
Thanks. Removed.

SeppoTakalo commented:
Fixed suggestions from @gchwier

rlubos reviewed (approved):
LGTM

jukkar reviewed (approved):
Sorry, I missed the earlier discussion. IMHO the various unit tests under `tests/` should be self contained and not have any outside connection requirements etc. so in that respect this could be placed under samples. But logically this fits better under tests/ and as it needs to be run manually anyway so I suppose we can go like this atm. Could be moved to some other place later if needed.

PerMac commented:
@jukkar "it has to be run manually" - currently, yes. But I believe those can be fully automated if a proper environment is available.  It can be discussed if we would like to expand CI testing with such configuration.

PerMac commented:
fyi @nashif
